in_toto/gpg/util.py

```diff
@@ -0,0 +1,222 @@
+  """
+  <Purpose>
+    Hash data prior to signature verification in conformance of the RFC4880
+    openPGP standard.
+
+  <Arguments>
+    headers: the additional OpenPGP headers as populated from
+    gpg_generate_signature
+
+    algorithm: The hash algorithm object defined by the cryptography.io hashes
+    module
+
+    content: the signed content
+
+  <Exceptions>
+    None
+
+  <Side Effects>
+    None
+
+  <Returns>
+    The RFC4880-compliant hashed buffer
+  """
+  # as per RFC4880 Section 5.2.2 paragraph 4, we need to hash the content,
+  # signature headers and add a very opinionated trailing header
+  hasher = hashing.Hash(algorithm, backend=backends.default_backend())
+  hasher.update(content)
+  hasher.update(headers)
+  hasher.update(b'\x04\xff')
+  hasher.update(struct.pack(">I", len(headers)))
+
+  return hasher.finalize()
+
+def parse_packet_header(data, expected_type):
+  """
+  <Purpose>
+    Parse an RFC4880 packet header and return its payload
+
+  <Arguments>
+    data: the packet header buffer
+
+    expected_type: The type of packet expected, as described in section 5.2.3.1
+        of RFC4880.
+
+  <Exceptions>
+    None
+
+  <Side Effects>
+    None
+
+  <Returns>
+    The RFC4880-compliant hashed buffer
+  """
+  data = bytearray(data)
+  packet_type = (data[0] & 0x3c ) >> 2
+  packet_length_bytes = data[0] & 0x03
+
+  ptr = 3
+  if packet_length_bytes == 1:
+    packet_length = struct.unpack(">H", data[1:ptr])[0]
+  else:
+    packet_length = data[1]
+    ptr = 2
+
+  if packet_type != expected_type: # pragma: no cover
+    raise in_toto.gpg.PacketParsingError("Expected packet {}, "
+        "but got {} instead!".format(expected_type, packet_type))
+
+  return data[ptr:ptr+packet_length]
+
+
+def compute_keyid(pubkey_packet_data):
+  """
+  <Purpose>
+    compute a keyid from an RFC4880 public-key buffer
+
+  <Arguments>
+    pubkey_packet_data: the public-key packet buffer
+
+  <Exceptions>
+    None
+
+  <Side Effects>
+    None
+
+  <Returns>
+    The RFC4880-compliant hashed buffer
+  """
+  hasher = hashing.Hash(hashing.SHA1(), backend=backends.default_backend())
+  hasher.update(b'\x99')
+  hasher.update(struct.pack(">H", len(pubkey_packet_data)))
+  hasher.update(bytes(pubkey_packet_data))
+  return binascii.hexlify(hasher.finalize())
+
+def parse_subpackets(subpacket_octets):
+  """
+  <Purpose>
+    parse the subpackets fields
+
+  <Arguments>
+    subpacket_octets: the unparsed subpacketoctets
+
+  <Exceptions>
+    in_toto.gpg.PacketParsingError if the octets are malformed
+
+  <Side Effects>
+    None
+
+  <Returns>
+    A list of tuples with like:
+        [ (packet_type, data),
+          (packet_type, data),
+          ...
+        ]
+  """
+  parsed_subpackets = []
+  ptr = 0
+  while ptr < len(subpacket_octets):
+    length = subpacket_octets[ptr]
+    ptr += 1
+    if length > 192 and length < 255 : # pragma: no cover
+      length = ((length - 192 << 8) + (subpacket_octets[ptr] + 102))
```

Comment:
I've seen a lot of these magic numbers now.  Consider using constants for these.

Reply:
I was thinking of doing this, but the RFC doesn't name them. I'll add a comment mentioning where the numbers come from.